### PR TITLE
fix(sync): prevent race condition by relying on autoincrement

### PR DESCRIPTION
**Pull request by max-nextcloud**
Prevent a possible race condition when two clients add steps at the same time.

See #4600.

Rely on the autoincrementing id in order to provide a canonical order that steps can be retrieved in.

When two clients push steps at the same time
the entries receive destinct ids that increment.
So if another client fetches steps in between
it will see the smaller id as the version of the fetched step and fetch the other step later on.

Transition:
In the future we can drop the version column entirely but currently there are still steps stored in the database that make use of the old column.
So we need to transition away from that.

In order to find entries that are newer than version x we select those that have both a version and an id larger than x.

Entries of the new format are newer than any entry of the old format. So we set their version to the largest possible value. This way they will always fulfill the version condition and the condition on the id is more strict and therefore effective.

For the old format the version will be smaller than the id as it's incremented per document while the id is unique accross documents. Therefore the version condition is the more strict one and effective.

The only scenario where the version might be larger than the id would be if there's very few documents in the database and they have had a lot of steps stored in single database entries.

### 🏁 Checklist

- [x] Code is properly formatted (`npm run lint` / `npm run stylelint` / `composer run cs:check`)
- [x] [Sign-off message](https://probot.github.io/apps/dco/) is added to all commits
- [ ] [Tests](https://github.com/nextcloud/text#-testing-the-app) (unit, integration and/or end-to-end) passing and the changes are covered with tests
- [x] Documentation is not required

### CODE REVIEW

**Comment by max-nextcloud** [comment]
Obviously WIP - opened the PR to be able to discuss the approach.

File: lib/Service/DocumentService.php [hunk 1]
```diff
@@ -263,15 +263,14 @@ private function insertSteps(Document $document, Session $session, array $steps,
...
-			$this->cache->set('document-version-' . $document->getId(), $newVersion);
+			$this->logger->debug("Adding steps to " . $document->getId() . ": bumping version from $stepsVersion");
 			$step = new Step();
 			$step->setData($stepsJson);
 			$step->setSessionId($session->getId());
 			$step->setDocumentId($document->getId());
-			$step->setVersion($newVersion);
-			$this->stepMapper->insert($step);
+			$step = $this->stepMapper->insert($step);
+			$newVersion = $step.id;
```
Comment (juliushaertl) [inline lib/Service/DocumentService.php:272]:
```suggestion
			$newVersion = $step->getId();
```

File: lib/Db/Step.php [hunk 2]
```diff
@@ -37,9 +37,19 @@
...
+	 * In order to make sure that new steps always
+	 * have a larger version than those that used the version field
+	 * use the largest possible value for big int
+	 */
+	public const VERSION_STORED_IN_ID = 18446744073709551615;
+
 	public $id = null;
 	protected string $data = '';
-	protected int $version = 0;
+	protected int $version = VERSION_STORED_IN_ID;
```
Comment (juliushaertl) [inline lib/Db/Step.php:52]:
```suggestion
	protected int $version = self::VERSION_STORED_IN_ID;
```

File: lib/Db/Step.php [hunk 3]
```diff
@@ -55,10 +65,13 @@ public function jsonSerialize(): array {
 		if (\json_last_error() !== JSON_ERROR_NONE) {
 			throw new \InvalidArgumentException('Failed to parse step data');
 		}
+		$version = $this->getVersion() == VERSION_STORED_IN_ID
```
Comment (juliushaertl) [inline lib/Db/Step.php:68]:
```suggestion
		$version = $this->getVersion() == self::VERSION_STORED_IN_ID
```

**Comment by mejo-** [comment]
I like the approach, it's a way cleaner solution than my approach to make read+write of `version` atomic. I pushed some fixes (force-pushed code-style fixes into the original commit and added another commit for logical fixes).

**Comment by mejo-** [comment]
Successfully tested the following scenario:


- main branch checked out
- client1 creates document and creates steps
- client2 joins document session and creates steps
- client2 leaves document session
- switch to PR branch
- client3 joins document session and creates steps
- client2 joins document session and creates steps
- client1 creates steps

All change are synced between all clients as expected.

**Review by mejo- — APPROVED** [review]
Ready for merge from my perspective.

**Comment by max-nextcloud** [comment]
There are two things that i did not look into yet that might be worthwhile ensuring:
* I'm not sure how the database queries perform now. We're now limiting by both `version` and `id` and I think there's an index on each of them but no combined index. Since these endpoints are called very often would be good to ensure there's no performance regression. Maybe we can get hold of the sql query that's run right now and run `EXPLAIN` on it to see which indexes it would use.
* There's a theoretical possibility that the highest version in the steps table is larger than the largest id. Versions can increment by more than one so if i only have one document that has a few steps and some of them increase the version by more than one the resulting version will be higher than the max id. This is unlikely but i am unsure if we can completely ignore it. It could be mitigated by inserting a step with a higher id that does not serve any other purpose. Not sure that cornercase is worth the effort though.

**Comment by mejo-** [comment]
> There's a theoretical possibility that the highest version in the steps table is larger than the largest id. Versions can increment by more than one so if i only have one document that has a few steps and some of them increase the version by more than one the resulting version will be higher than the max id. This is unlikely but i am unsure if we can completely ignore it. It could be mitigated by inserting a step with a higher id that does not serve any other purpose. Not sure that cornercase is worth the effort though.

I thought a bit about this and I think we should ignore this corner case. Nextcloud instances with just one (or very few) document sessions likely aren't used collaboratively. Sure, there's a theoretical possibility that people setup a Nextcloud instance to work together on one file, but that's not very likely. And even in this scenario, the autosave should have saved the last document state. So worst case in this very unlikely scenario people will loose a few editing steps. This is not worth adding complex migration logic in my opinion.

**Comment by mejo-** [comment]
>     I'm not sure how the database queries perform now. We're now limiting by both version and id and I think there's an index on each of them but no combined index. Since these endpoints are called very often would be good to ensure there's no performance regression. Maybe we can get hold of the sql query that's run right now and run EXPLAIN on it to see which indexes it would use.

@juliushaertl and me looked into this. The index on `document_id` is always used first and the following where statements are only executed on its subset. In our tests, the new queries didn't look worse regarding performance than the old ones.

**Comment by mejo-** [comment]
/backport ea208ee0d69f08241f7525352e9affd060ff34e9,308d64069f8903a113985d7d71d027b127a3b0ce to stable27

**Pull request merged**